### DISCUSSION TITLE
Lots of missing files?

### DISCUSSION THREAD

**Discussion by jgng** [ReagentX/imessage-exporter#400] [Q&A] [answered]
I first tried using my MacOS backup, but I noticed in my first export that I got lots of "attachment does not exist!" errors in my html files. I had a hunch that maybe my MacOS device (which is new) didn't have my full message history downloaded from iCloud yet, so I switched to using an iOS unencrypted backup, following your instructions to point to the path. Same issue. I ran in diagnostic mode and got the following:

```
iMessage Database Diagnostics

Message diagnostic data:
    Total messages: 91451
    Messages not associated with a chat: 8
    Messages belonging to more than one chat: 19
Attachment diagnostic data:
    Total attachments: 5579
        Data referenced in table: 17.57 GB
        Data present on disk: 10.66 GB
    Missing files: 1833 (33%)
        No path provided: 2
        No file located: 1831
Global diagnostic data:
    Total database size: 8.22 KB
    Duplicated contacts: 491
    Duplicated chats: 95
```

What would cause my database to be missing 33% of files? I understand this is probably not an issue with imessage-exporter but I'm hoping someone here can help me resolve this. My iMessages are backed up in iCloud -- maybe they weren't all downloaded to my device at the time of the backup? Is that something I can change? Or alternately, is there a way to get them all downloaded on my Mac OS system and run the export from that DB?

For reference, here's what I see in diagnostic mode using my MacOS chat database. Seems like there are many fewer attachments total…

```
iMessage Database Diagnostics

Message diagnostic data:
    Total messages: 91462
    Messages not associated with a chat: 3938
Attachment diagnostic data:
    Total attachments: 4207
        Data referenced in table: 14.80 GB
        Data present on disk: 12.43 GB
    Missing files: 120 (3%)
        No path provided: 120
        No file located: 0
Global diagnostic data:
    Total database size: 130.61 MB
    Duplicated contacts: 491
    Duplicated chats: 80
```

**Comment by ReagentX** [comment, marked as answer]
The difference between your macOS and iOS diagnostic data suggests an issue with iCloud sync, sadly there isn't much we can do to fix that. In the iMessage app, if you click the info button for a conversation and scroll down, there is a button to force it to download all of the attachments for that conversation:

<img width="321" alt="image" src="https://github.com/user-attachments/assets/8fd8e7a2-1d8c-46a5-9d8e-8645317e2458">

On iOS specifically, Messages is probably more aggressive in not taking up storage with old attachments than macOS.

All that said, the issue you are experiencing with "Attachment does not exist!" in the HTML output indicates that you have rows in the `message` table that reference attachments, but there is no matching attachment stored in the `attachment` table to insert into those messages.
